bundler/lib/bundler/checksum.rb

```diff
@@ -2,6 +2,71 @@
 
 module Bundler
   class Checksum
+    class << self
+      def from_file_source(file_source, digest_algorithms: %w[sha256])
+        raise ArgumentError, "not a valid file source: #{file_source}" unless file_source.respond_to?(:with_read_io)
+
+        digests = digest_algorithms.map do |digest_algorithm|
+          [digest_algorithm.to_s, Bundler::SharedHelpers.digest(digest_algorithm.upcase).new]
+        end.to_h
+
+        file_source.with_read_io do |io|
+          until io.eof?
+            block = io.read(16_384)
+            digests.each_value {|digest| digest << block }
+          end
+
+          io.rewind
+        end
+
+        digests.map do |algo, digest|
+          Checksum.new(algo, digest.hexdigest!, file_source.path)
+        end
+      end
+    end
+
+    attr_reader :algo, :digest, :sources
+    def initialize(algo, digest, source)
+      @algo = algo
+      @digest = digest
+      @sources = [source]
+    end
+
+    def ==(other)
+      match?(other) && other.sources == sources
+    end
+
+    alias_method :eql?, :==
+
+    def match?(other)
+      other.is_a?(self.class) && other.digest == digest && other.algo == algo
+    end
+
+    def hash
+      digest.hash
+    end
+
+    def to_s
+      "#{to_lock} (from #{sources.first}#{", ..." if sources.size > 1})"
+    end
+
+    def to_lock
+      "#{algo}-#{digest}"
```

Comment:
What about to use format used by [Digest Fields](https://datatracker.ietf.org/doc/draft-ietf-httpbis-digest-headers/, https://www.rfc-editor.org/rfc/rfc8941.html#name-serializing-a-byte-sequence)? I know this is not directly related to HTTP protocol, but it could make it easier later to re-use.

Or at least unify algo naming (use `sha-256` insetad of `sha256`) as use at https://github.com/rubygems/rubygems/compare/master...opaque-etag-http-digest#diff-603a4693e8a95163d56f6e250d52ec0c280378829f2f81c967385c38ca8a6a64R9.

Reply:
I like the idea of using something standardized. The first few links suggest `":#{base64}:"`. We currently use most of our checksums as hexdigests, but there's code to accept either from the API.

Also, looking at your last link I see there's this [existing error class](https://github.com/rubygems/rubygems/compare/master...opaque-etag-http-digest#diff-603a4693e8a95163d56f6e250d52ec0c280378829f2f81c967385c38ca8a6a64R11-R22) for checksum mismatches.

Reply:
Reading into it, I'm thinking about: `sha-256=FEEDDADC0FFEE111DECAF15BAD`

Hex is slightly easier, though it takes a few extra bytes. The command line output of shasum or openssl can output hex by default instead of having to pass it through as binary to `openssl base64`. I don't feel strongly here though, because base64 is more compact and it's not exactly uncommon. I'd be happy to convert or even support both out of the box (easy because string size differs)

For the algorithm, you're definitely right that "sha-256" is the more common formatting. We use `Digest()` in that code you linked, so maybe it can find the class even with the dash there. I'll try.

Reply:
> We use Digest() in that code you linked, so maybe it can find the class even with the dash there. I'll try.

sadly it can't :(

Reply:
Feel free to extract that one into some generic namespace and use, I'll update my branch to use the same.

Reply:
Well, [all](https://ruby-doc.org/stdlib-3.0.1/libdoc/digest/rdoc/Digest.html) of [ruby](https://ruby-doc.org/stdlib-3.0.1/libdoc/digest/rdoc/Digest.html) seems to use `SHA256` without the dash. Since we're using ruby, I'll just change to an `=` and call it good. Some future algorithm may want to use the dash (e.g. "SHA3-256")

Reply:
I think they mean different things. One is used to retry partial downloads of compact index when Etag doesn't match. The other is a security error. I'll keep them separate.